var/spack/repos/builtin/packages/rsbench/package.py

```diff
@@ -0,0 +1,62 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+from spack import *
+
+
+class Rsbench(MakefilePackage):
+    """A mini-app to represent the multipole resonance representation lookup
+       cross section algorithm."""
+
+    homepage = "https://github.com/ANL-CESAR/RSBench"
+    url = "https://github.com/ANL-CESAR/RSBench/archive/v2.tar.gz"
+
+    variant('pgi', default=False, description='Build with PGI.')
+
+    depends_on('pgi', when='+pgi')
+
+    version('2', '15a3ac5ea72529ac1ed9ed016ee68b4f')
+    version('0', '3427634dc5e7cd904d88f9955b371757')
+
+
+    @property
+    def build_targets(self):
+
+        targets = [
+            '--directory=src'
+        ]
+
+        if '%intel' in self.spec:
+            targets.append('COMPILER=intel')
+
+        if '%pgi' in self.spec:
+            targets.append('COMPILER=pgi')
+            targets.append('CC={0}'.format(self.spec['pgi'].pgcc))
```

Comment:
@pkondamudi Here are my thoughts:

1. Are you packaging your own software, or software that you contribute to? Most people don't have the luxury of designing their Makefile, they have to hack things to make them work. If you do have the power to control your Makefile, the best option is to provide default settings for `CC` and `CFLAGS`, but always let the user override them on the command line or through environment variables. I like using `?=` as described [here](https://stackoverflow.com/a/448939/5828163). `?=` respects environment variables, but `=` and `:=` ignore them. All 3 will respect variables set on the command line.

2. First of all, there is no such thing as `spec['pgi'].pgcc`. Packages shouldn't add a dependency on any particular compiler (most build with a compiler, and can be built with most compilers). Therefore, `pgi` won't be in `spec` anyway. Also, `spec['pgi']` has no `pgcc` attribute. We added `mpicc` and friends to each MPI package because we needed a generic way of referring to them that works no matter what the package decides to name it. If your package builds with MPI, you should use `spec['mpi'].mpicc`. If not, then see below.

3. I guess I don't understand how this is different from technique 1?

4. If your Makefile is well designed, it will honor environment variables like `CC` and `CFLAGS`. If it is poorly designed, these environment variables will be overridden and you will have to use one of the techniques listed below. 

> how does it choose which compiler to set

Spack chooses which compiler to set based on what you tell it to use. If you install with:
```
$ spack install rsbench %intel
```
it will use the Intel compilers. If you don't specify, Spack will look in `packages.yaml` to see what your default compiler is. If you don't override this, it is `[gcc, intel, pgi, clang, xl, nag]` on Linux and `[clang, gcc, intel]` on macOS. Spack will find the first compiler in this list that is present in your `packages.yaml` and use that compiler.

There are a couple other things that you should understand. First of all, one of the main tenants of Spack is that it uses compiler wrappers instead of calling the compilers you specify directly. These can be found in `lib/spack/env`. The compiler wrapper is called `cc`, and we have symlinks for other common compiler names in the same directory. Regardless of what you're doing, it is imperative that these compiler wrappers are called instead of the actual compilers. These wrappers are responsible for things like automatically adding `-L`, `-I`, and `-Wl,-rpath` for all of your package's dependencies.

Inside a `package.py`, you can refer to the absolute path to the compiler wrappers with `spack_cc`. The absolute path to the actual compiler is `self.compiler.cc`, but don't ever use that. The compiler wrappers will wrap around the real compiler and inject these flags for you. If you use a compiler name like `cc` instead of the absolute path, you should still be safe, as `build_environment.py` ensures that this directory is prepended to the front of your `PATH`.

Let me know if any of this is unclear or you have any other questions about compilers!